File: src/wings/query/engine.py
"""Query engine — unified LLM API entry point.

Looks up the provider via ModelRegistry and calls it. Handles retry
for transient errors. Does NOT own message format conversion (that's
the provider's job) or model selection (that's ModelSelector).
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any, AsyncIterator

from wings.messages.types import Message, StreamEvent
from wings.models.protocol import ModelConfig, ModelResponse
from wings.models.registry import ModelRegistry

logger = logging.getLogger(__name__)

# Transient errors worth retrying
_RETRIABLE_STATUSES: tuple[int, ...] = (429, 500, 502, 503, 504)


class QueryError(Exception):
    """A non-retriable error from a model provider."""


class QueryEngine:
    """Unified entry point for LLM API calls.

    Messages are passed through to the provider as-is — each provider
    converts to its native format internally.  The engine is only
    responsible for error handling and retry logic.
    """

    def __init__(
        self,
        registry: ModelRegistry,
        *,
        max_retries: int = 3,
        retry_base_delay: float = 1.0,
    ):
        self._registry = registry
        self._max_retries = max_retries
        self._retry_base_delay = retry_base_delay

    async def chat(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        config: ModelConfig,
    ) -> ModelResponse:
        """Non-streaming query. Returns a complete ModelResponse."""
        provider = self._registry.get(model)

        last_error: Exception | None = None
        for attempt in range(self._max_retries + 1):
            try:
                return await provider.chat(messages, tools, config)
            except Exception as e:
                last_error = e
                if not self._is_retriable(e) or attempt == self._max_retries:
                    break
                delay = self._retry_base_delay * (2 ** attempt)
                logger.warning(
                    "retry %d/%d for %s after %.1fs: %s",
                    attempt + 1, self._max_retries, model, delay, e,
                )
                await asyncio.sleep(delay)

        raise QueryError(f"chat failed for {model}: {last_error}") from last_error

    async def stream(
        self,
        messages: list[Message],
        model: str,
        tools: list[dict[str, Any]] | None,
        config: ModelConfig,
    ) -> AsyncIterator[StreamEvent]:
        """Streaming query. Yields StreamEvent items."""
        provider = self._registry.get(model)

        last_error: Exception | None = None
        for attempt in range(self._max_retries + 1):
            try:
                async for event in provider.stream(messages, tools, config):
                    yield event
                return  # successful completion
            except Exception as e:
                last_error = e
                if not self._is_retriable(e) or attempt == self._max_retries:
                    break
                delay = self._retry_base_delay * (2 ** attempt)
                logger.warning(
                    "stream retry %d/%d for %s after %.1fs: %s",
                    attempt + 1, self._max_retries, model, delay, e,
                )
                await asyncio.sleep(delay)

        raise QueryError(f"stream failed for {model}: {last_error}") from last_error

    def _is_retriable(self, exc: Exception) -> bool:
        """Check whether an exception suggests a transient error."""
        status = getattr(exc, "status_code", None)
        if status is not None and status in _RETRIABLE_STATUSES:
            return True
        response = getattr(exc, "response", None)
        if response is not None:
            http_status = getattr(response, "status_code", None)
            if http_status is not None and http_status in _RETRIABLE_STATUSES:
                return True
        import httpx
        if isinstance(exc, httpx.HTTPStatusError):
            return exc.response.status_code in _RETRIABLE_STATUSES
        if isinstance(exc, (httpx.ConnectError, httpx.ReadError, httpx.RemoteProtocolError)):
            return True
        return False

File: src/query/engine.ts
/**
 * Query engine — unified LLM API entry point.
 *
 * Looks up the provider via ModelRegistry and calls it. Handles retry
 * for transient errors. Does NOT own message format conversion (that's
 * the provider's job) or model selection (that's ModelSelector).
 */

import type { Message, MessageContent, StreamEvent } from "../messages/types.ts";
import type { ModelConfig, ModelProvider, ModelResponse } from "../models/protocol.ts";
import type { ModelRegistry } from "../models/registry.ts";

// Transient errors worth retrying
const RETRIABLE_STATUSES = new Set([429, 500, 502, 503, 504]);

export class QueryError extends Error {
  constructor(message: string, public cause?: unknown) {
    super(message);
    this.name = "QueryError";
  }
}

export class QueryEngine {
  constructor(
    private _registry: ModelRegistry,
    private _maxRetries: number = 3,
    private _retryBaseDelay: number = 1.0,
  ) {}

  async chat(
    messages: Message[],
    model: string,
    tools: Record<string, unknown>[] | null,
    config: ModelConfig,
  ): Promise<ModelResponse> {
    const provider = this._registry.get(model);

    let lastError: unknown = null;
    for (let attempt = 0; attempt <= this._maxRetries; attempt++) {
      try {
        return await provider.chat(messages, tools, config);
      } catch (e) {
        lastError = e;
        if (!this._isRetriable(e) || attempt === this._maxRetries) break;
        const delay = this._retryBaseDelay * Math.pow(2, attempt);
        console.warn(`retry ${attempt + 1}/${this._maxRetries} for ${model} after ${delay.toFixed(1)}s: ${(e as Error).message}`);
        await sleep(delay * 1000);
      }
    }
    throw new QueryError(`chat failed for ${model}: ${(lastError as Error)?.message ?? lastError}`, lastError);
  }

  async *stream(
    messages: Message[],
    model: string,
    tools: Record<string, unknown>[] | null,
    config: ModelConfig,
  ): AsyncGenerator<StreamEvent | MessageContent> {
    const provider = this._registry.get(model);

    let lastError: unknown = null;
    for (let attempt = 0; attempt <= this._maxRetries; attempt++) {
      try {
        for await (const event of provider.stream(messages, tools, config)) {
          yield event;
        }
        return; // successful completion
      } catch (e) {
        lastError = e;
        if (!this._isRetriable(e) || attempt === this._maxRetries) break;
        const delay = this._retryBaseDelay * Math.pow(2, attempt);
        console.warn(`stream retry ${attempt + 1}/${this._maxRetries} for ${model} after ${delay.toFixed(1)}s: ${(e as Error).message}`);
        await sleep(delay * 1000);
      }
    }
    throw new QueryError(`stream failed for ${model}: ${(lastError as Error)?.message ?? lastError}`, lastError);
  }

  private _isRetriable(exc: unknown): boolean {
    const e = exc as any;
    // Anthropic/OpenAI SDK errors use `status`.
    const status = e?.status_code ?? e?.status;
    if (typeof status === "number" && RETRIABLE_STATUSES.has(status)) return true;
    const response = e?.response;
    const httpStatus = response?.status_code ?? response?.status;
    if (typeof httpStatus === "number" && RETRIABLE_STATUSES.has(httpStatus)) return true;
    // fetch() throws TypeError on network/connection failures.
    if (e instanceof TypeError) return true;
    return false;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
